pkg/tenantctl/controller.go

```diff
@@ -157,219 +172,470 @@ func (c *Controller) processNextWorkItem() bool {
+	reconcileErr := c.reconcile(config, tenant)
 
-	// Check if tenant setup is completed
-	if tenant.Status.Progress != api.TenantProgressFinished {
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressInProcess)
-		var err error
-		var namespaceName string
-		var account *k8s.ServiceAccountWrap
-
-		config, err := getClientConfig(c.factory, tenant.GetNamespace())
-		if err != nil {
-			log.Printf("ERROR: Could not get config: %s", err.Error())
+	// do not update the status if there's no change
+	if !equality.Semantic.DeepEqual(origTenant.Status, tenant.Status) {
+		if _, err := c.updateStatus(tenant); err != nil {
+			if !c.isInitialized(origTenant) && c.isInitialized(tenant) {
+				c.rollbackTenantNamespace(tenant.Status.TenantNamespaceName, tenant, config)
+			}
 			return err
 		}
-		tenantRoleName := config.GetTenantRoleName()
+	}
 
-		//TODO: handle updateProgress errors
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressCreateNamespace)
-		namespaceName, err = c.createNamespace(tenant)
-		if err != nil {
-			return c.handleError(tenant, err, api.TenantResultErrorContent)
-		}
-		log.Printf("Create namespace successful for %s", namespaceName)
+	if reconcileErr != nil {
+		return reconcileErr
+	}
 
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressGetServiceAccount)
-		account, err = c.getServiceAccount(tenant, defaultServiceAccountName)
-		if err != nil {
-			return c.handleError(tenant, err, api.TenantResultErrorInfra)
-		}
+	c.updateMetrics()
+	c.syncCount++
+	return nil
+}
 
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressAddRoleBinding)
-		var roleBinding *v1beta1.RoleBinding
-		roleBinding, err = c.addRoleBinding(account, tenant, tenantRoleName)
-		if err != nil {
-			return c.handleError(tenant, err, api.TenantResultErrorInfra)
-		}
-		log.Printf("Created Role Binding '%s' in namespace '%s'", roleBinding.GetName(), namespaceName)
+func (c *Controller) isInitialized(tenant *api.Tenant) bool {
+	return tenant.Status.TenantNamespaceName != ""
+}
 
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressFinalize)
-		tenant.Status.Result = api.TenantResultSuccess
-		tenant.Status.Message = "Tenant namespace successfully prepared"
-		if tenant, err = c.updateStatus(tenant); err != nil {
-			return err
-		}
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressFinished)
-		log.Printf("Tenant preparation successful for %s", tenant.GetName())
+func (c *Controller) reconcile(config clientConfig, tenant *api.Tenant) (err error) {
+	if c.isInitialized(tenant) {
+		err = c.reconcileInitialized(config, tenant)
+	} else {
+		err = c.reconcileUninitialized(config, tenant)
 	}
-	c.updateMetrics()
-	c.syncCount++
+	return
+}
+
+func (c *Controller) reconcileUninitialized(config clientConfig, tenant *api.Tenant) error {
+	c.logPrintln(tenant, "tenant not initialized yet")
+
+	nsName, err := c.createTenantNamespace(config, tenant)
+	if err != nil {
+		condMsg := fmt.Sprintf("Failed to create the tenant namespace.")
```

Comment:
I agree that we might not want to expose the real error message to the client.
But we should maybe log the error with an ID and mention this ID in the message. Otherwise we might not be able to identify the correct issue.

Simpler would be to really add the original error message. Could this contain any confidential internals?

Reply:
The message in the condition is user-facing (might even be shown to the end user via the client). I don't want users to see technical error messages.

An operator can look into the controller log (or later into the event log of the affected resource) to see the technical error message. If the error persists, respective log entries should be created periodically.

If we want to have error ids, we need to add some supporting infrastructure code. That's to much for this change.

Reply:
Correction: Clients should not show such message to their end users.

However, I think the message in the condition should only refer to concepts from the definition of the resource type. Underlying causes, implementation details and so on should not appear here but in the log.